File: src/core/errors/DomainError.ts
export class DomainError extends Error {
  constructor(message: string, public readonly code: string = 'DOMAIN_ERROR') {
    super(message);
    this.name = 'DomainError';
  }
}

export class DuplicateActionError extends DomainError {
  constructor(message: string = 'Ushbu amal avval bajarilgan.') {
    super(message, 'DUPLICATE_ACTION');
    this.name = 'DuplicateActionError';
  }
}

export class EntityNotFoundError extends DomainError {
  constructor(entityName: string, id: string) {
    super(`${entityName} topilmadi: ${id}`, 'NOT_FOUND');
    this.name = 'EntityNotFoundError';
  }
}

File: mobile/lib/core/errors/domain_error.dart
class DomainError implements Exception {
  final String message;
  final String code;

  const DomainError(this.message, {this.code = 'DOMAIN_ERROR'});

  @override
  String toString() => 'DomainError: $message (code: $code)';
}

class DuplicateActionError extends DomainError {
  const DuplicateActionError([super.message = 'Ushbu amal avval bajarilgan.'])
      : super(code: 'DUPLICATE_ACTION');
}

class EntityNotFoundError extends DomainError {
  final String entityName;
  final String id;

  EntityNotFoundError(this.entityName, this.id)
      : super('$entityName topilmadi: $id', code: 'NOT_FOUND');
}
